Update network ID to chain ID in the protocol (#167)
LiskHQ/lips: proposals/lip-0037.md

```
LIP: 0037
Title: Use message tags and chain identifiers for signatures
Author: Andreas Kendziorra <andreas.kendziorra@lightcurve.io>
Discussions-To: https://research.lisk.com/t/use-message-tags-and-network-identifiers-for-signatures/

> Status: Draft

Reviewer: Should we add that this LIP replaces LIP 9?

Author: Is there a standard way to do this in the header?

Reviewer: See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0001.md#lip-header-preamble

Author: added

Chain identifiers are 4-byte values that follow a specific format: the first byte is used to identify the network in which the chain is running (either the mainnet, official testnet, or any other test network); the other 3 bytes identify the chain within the network. We include the network-specific prefix explicitly to ensure that a chain does not use the same chain identifier in the test network as in the mainnet.

## Specification

### Message Tags

Every binary message must be tagged before being signed. This is done as specified by the function `tagMessage` in the pseudo code below: A specific ASCII-encoded tag and a chain identifier as specified in [LIP 0009](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0009.md#specification) are prepended to the message.

```python
def tagMessage(tag: bytes, chainID: bytes, message: bytes) -> bytes:

>     return tag + networkIdentifier + message

Reviewer 2: `tag + chainID + message`

### Message Tags

Prepending a tag to a binary message before signing, where the tag is specific to the type of the message, prevents the signature re-usage in general. Consider again the blockheader `bh` and the transaction `tx` that are serialized to the same binary message `bm`. When the tag `"LSK_BH_"` is used for block header signatures and the tag `"LSK_TX_"` for transaction signatures, then the signatures for `bh` and `tx` are computed by signing the messages `"LSK_BH_" + bm` and `"LSK_TX_" + bm` respectively. Hence, the signature of `bh` cannot be a valid signature for `tx` and vice versa (assuming no collisions in the signing function).

Note that any updates to the serialization specification of a message type must be accompanied by introducing a new tag. For example, when the [transaction schema](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0028.md#transactionschema-schema) is updated, a new tag like `"LSK_TX:V2_"` must be defined and used.

### Chain Identifiers

> Chain identifiers for transaction signatures and block signatures were already introduced in [LIP 0009](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0009.md#specification) and [LIP 0024](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0024.md#update-to-the-block-header-signing-procedure) (as network identifiers) to prevent replay attacks on other chains. Here, we specify how to use a chain identifier for any kind of message to prevent replay attacks on other chains in general.

Reviewer: I would write here explicitly that the specifications in this LIP supersede those in LIP 9. The way in which we use chain ID for signing is the same as we used network identfiers, however, the length and how they are obtained is now different. I think it would be nice to also elaborate on the reasoning, for instance, something like the following:

> For users to be able to easily verify that they sign a transaction for the correct blockchain, we choose to abandon the usage of 32-byte network identifiers that are computed as a hash value from the genesis block ID and community identifier. Instead, the new chain identifier which also serve the purpose to prevent transaction replay among different blockchains, has a length of only 4 bytes. It further can be chosen by the blockchain creator. For the official mainnet blockchain of a project, we always recommend to register the identifier on the Lisk Mainchain to ensure uniqueness in the Lisk ecosystem.

Author: added something in this direction

Reviewer: I like the new formulation :-)

Chain identifiers are 4-byte values that follow a specific format: the first byte is used to identify the network in which the chain is running (either the mainnet, official testnet, or any other test network); the other 3 bytes identify the chain within the network. We include the network-specific prefix explicitly to ensure that a chain does not use the same chain identifier in the test network as in the mainnet.

## Specification

### Message Tags

> Every binary message must be tagged before being signed. This is done as specified by the function `tagMessage` in the pseudo code below: A specific ASCII-encoded tag and a chain identifier as specified in [LIP 0009](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0009.md#specification) are prepended to the message.

Reviewer: The chain identifier should be specified in this LIP and is different compared to the network identifier in LIP 9 which is defined as `network identifier = H(nethash + community identifier)`.

Author: Do you think I have to write this explicitly? It is somehow obvious that here we are given the new specs

Reviewer: ```suggestion
Every binary message must be tagged before being signed. This is done as specified by the function `tagMessage` in the pseudo code below: A specific ASCII-encoded tag and a chain identifier are prepended to the message.
```

Reviewer: I just wanted to say the reference to LIP 9 here is rather confusing then helpful so I would just remove it.

When a new tag is required in the future, the format:

    "LSK_" + TAG + "_"

has to be used where:

- Strings in double quotes are ASCII-encoded literals.
- TAG is a unique string indicating the scheme and, optionally, additional information, e.g., to specify a version number as shown [above](#message-tags). TAG must contain only ASCII characters between 0x21 and 0x7e (inclusive), except that it must not contain underscore (0x5f).

### Chain Identifiers

> Chain identifiers are 4-byte values. The first byte is set to `CHAIN_ID_PREFIX_MAINNET` for chains running in the mainnet network and to `CHAIN_ID_PREFIX_TESTNET` for chains running in the testnet network. The other 3 bytes are unique to that chain.

Reviewer: ```suggestion
Chain identifiers are 4-byte values. The first byte is set to `CHAIN_ID_PREFIX_MAINNET` for chains running in the mainnet network and to `CHAIN_ID_PREFIX_TESTNET` for chains running in the testnet network. The other 3 bytes must be chosen uniquely for the respective blockchain, i.e., no other blockchain created with the Lisk SDK should use the same 3 bytes.
```

Reviewer: To clarify what means unique here.

The following table defines the chain-identifiers prefixes currently specified.

| Name                            | Type  | Value| Description                                                 |
|---------------------------------|-------|------|-------------------------------------------------------------|

> | `CHAIN_ID_PREFIX_MAINNET`       | bytes | 0x00 | Chain-identifiers prefix for chains running in the mainnet. |

Reviewer: ```suggestion
| `CHAIN_ID_PREFIX_MAINNET`       | bytes | 0x00 | Chain-identifier prefix for mainnet blockchains. |
```

Reviewer: "Running in the mainnet" sounds odd.

Reviewer: Do we mention here that it will be a requirement to have the prefix `0x00` to connect to Lisk Mainnet for interoperability (I am assuming there that we would make this a requirement) and reference to the chain registration LIP?

Author: I thought that this LIP should be unaware of other modules using the chainID for something else. Because otherwise we would also need to add the requirement that all token IDs in the mainnet start with that prefix etc..

> | `CHAIN_ID_PREFIX_TESTNET`       | bytes | 0x01 | Chain-identifiers prefix for chains running in the testnet. |

Reviewer: ```suggestion
| `CHAIN_ID_PREFIX_TESTNET`       | bytes | 0x01 | Chain-identifier prefix for testnet blockchains. |
```

| `CHAIN_ID_PREFIX_TESTNET`       | bytes | 0x01 | Chain-identifiers prefix for chains running in the testnet. |

### Signing and Verifying with Ed25519

> Let [`Sign`](https://tools.ietf.org/html/rfc8032#section-5.1.6) and [`Verify`](https://tools.ietf.org/html/rfc8032#section-5.1.7) be the signing and verifying functions of Ed25519 as specified in [RFC 8032](https://tools.ietf.org/html/rfc8032). Then the signature for a binary message `m` and a secret key `sk` is generated by `signMessage(sk, tag, networkIdentifier, m)` as defined below. `tag` must be the correct message tag for `m` and `networkIdentifier` the correct network identifier for the chain. The resulting signature `sig` in combination with the message `m` and the matching public key `pk` is verified by `verifyMessageSig(pk, tag, networkIdentifier, m, sig)`.

Reviewer 2: The very end of the paragraph:
`pk` is verified by `verifyMessageSig(pk, tag, networkIdentifier, m, sig)`. -> `pk` is verified by `verifyMessageSig(pk, tag, chainID, m, sig)`.
